### ISSUE TITLE
postgresql_configuration sync issue

### ISSUE THREAD

**Issue by ulucinar** [crossplane-contrib/provider-jet-azure#97]
### What happened?

Using the example manifests in the repo for postgresql server and postgresql server configuration, postgresql server object is successfully provisioned, becomes ready & synced and is stable, i.e., it stays in ready & synced state.

However, although initially postgresql configuration object becomes ready & synched, after a while it transitions to the ready but not-synced state because if the following error in `ExternalClient.Observe`:
```yaml
    message: 'observe failed: cannot run plan: plan failed: Instance cannot be destroyed:
      Resource azurerm_postgresql_configuration.example-psql-configuration has lifecycle.prevent_destroy
      set, but the plan calls for this resource to be destroyed. To avoid this error
      and continue with the plan, either disable lifecycle.prevent_destroy or reduce
      the scope of the plan using the -target flag.: File name: main.tf.json'
    reason: ReconcileError
    status: "False"
    type: Synced
```

### How can we reproduce it?

This can be reproduced using the example manifests mentioned above.

**Comment by muvaf** [comment]
> Instance cannot be destroyed

It seems like Terraform execution tried to delete this resource, which is probably because it detected a change in a field that is immutable. I think what we need to figure out is why suddenly it thought that a diff on that field exists without a user action.

**Comment by ulucinar** [comment]
Looks like this is related to the intricacies involved with these server configuration resources. For example, for the initial attempt (i.e., no configuration resource has initially been provisioned), the following manifest is successfully applied with `terraform apply`:
```
{
  "provider": {
    "azurerm": {
      "features": {}
    }
  },
  "resource": {
    "azurerm_postgresql_configuration": {
      "example-psql-configuration": {
        "lifecycle": {
          "prevent_destroy": true
        },
        "name": "array_nulls",
        "resource_group_name": "alper",
        "server_name": "example-psqlserver-alper",
        "value": "on"
      }
    }
  },
  "terraform": {
    "required_providers": {
      "azurerm": {
        "source": "hashicorp/azurerm",
        "version": "2.78.0"
      }
    }
  }
}
```

Note the `prevent_destroy` meta-arg. After this initial run, if you just change the configuration's value to `off` and rerun the same `terraform apply` command, it fails because of the lifecycle meta arg.

I did not check but it looks like it's implemented via a destroy. Btw, the Go SDK does not have a delete call but a createOrUpdate call AFAIR.

**Comment by muvaf** [comment]
It seems like there are two cases here:
* User changes `value` and terraform tries to re-create it.
  * Because that field is marked as `ForceNew`, i.e. immutable in Crossplane world: https://github.com/hashicorp/terraform-provider-azurerm/blob/3eefe84/internal/services/postgres/postgresql_configuration_resource.go#L54
* User does nothing and suddenly starts to see that error.
  * This could be related to late-init or plan functionality that causes `IsUpToDate` to return false negatives, which is worth looking into.

@ulucinar Is it the second one or the first one that you're seeking to solve by making `lifecycle.prevent_destroy` optional?

**Comment by ulucinar** [comment]
To clarify the above discussion, it's related to updates. For instance, if you are provisioning a configuration resource whose value (`spec.forProvider.value`) is different than the default value (or whatever the value currently set at server side).

**Comment by ulucinar** [comment]
What we are dealing with here is related to the first one. The spec parameters do not change (due to late-init):

```yaml
spec:
  deletionPolicy: Delete
  forProvider:
    resourceGroupName: alper
    resourceGroupNameRef:
      name: example
    serverName: example-psqlserver-alper
    serverNameRef:
      name: example-psqlserver-alper
    value: "off"
  providerConfigRef:
    name: example
status:
  atProvider: {}
  conditions:
  - lastTransitionTime: "2021-11-24T19:01:54Z"
    reason: Available
    status: "True"
    type: Ready
  - lastTransitionTime: "2021-11-24T19:01:54Z"
    reason: ReconcileSuccess
    status: "True"
    type: Synced
```

and then we get:

```yaml
spec:
  deletionPolicy: Delete
  forProvider:
    resourceGroupName: alper
    resourceGroupNameRef:
      name: example
    serverName: example-psqlserver-alper
    serverNameRef:
      name: example-psqlserver-alper
    value: "off"
  providerConfigRef:
    name: example
status:
  atProvider: {}
  conditions:
  - lastTransitionTime: "2021-11-24T19:01:54Z"
    reason: Available
    status: "True"
    type: Ready
  - lastTransitionTime: "2021-11-24T19:02:13Z"
    message: 'observe failed: cannot run plan: plan failed: Instance cannot be destroyed:
      Resource azurerm_postgresql_configuration.example-psql-configuration has lifecycle.prevent_destroy
      set, but the plan calls for this resource to be destroyed. To avoid this error
      and continue with the plan, either disable lifecycle.prevent_destroy or reduce
      the scope of the plan using the -target flag.: File name: main.tf.json'
    reason: ReconcileError
    status: "False"
    type: Synced
```

In the above example, the server-side configuration holds an `on`, and we have `off` in the `spec`, thus an update is needed. 

What confuses me, is why we declare it as `Ready` in the first hand. But regardless of this, the `prevent_destroy` seems to be blocking proper updates for this resource.

**Comment by muvaf** [comment]
> observe failed: cannot run plan: plan failed:

Oh I just noticed that it doesn't try to apply, it just tries to run `terraform plan`.

> After this initial run, if you just change the configuration's value to off and rerun the same terraform apply command, it fails because of the lifecycle meta arg.

@ulucinar have you tried `terraform plan` with those files in your local? I wonder if it returns that error even if there is no diff.

**Comment by muvaf** [comment]
@ulucinar let me know that the problem is that the first ever call we issue to check existence is `terraform apply -refresh-only`. However, there is no case that this resource does not exist, it always imports the server-side default, then you get the problem that there needs to be re-creation because resource already exists. So, any sort of creation is actually _re-creation_ for that resource.

Given the rarity of the case, I tried to think of ways to handle this behavior without exposing a top-level configuration toggle but I couldn't really find a way. There is simply no way we can know that the resource does not exist without calling `Refresh`, which causes import of the existing default config.

**Comment by turkenh** [comment]
**TL;DR**

Based on my experiments, the following should already be possible:

a. server side config=x -> Crossplane MR wants x -> ready&synced
b. server side config=x -> Crossplane MR wants y -> properly updated server-side to y -> ready&synced

What will NOT be possible is:

a. server-side config=x -> Crossplane MR wants y -> properly updated server-side to y -> ready&synced -> Crossplane MR updated and wants z -> it fails. 

I believe, without any changes, we could just say if you want to update it to z, you first need to delete MR with y and create a new MR with z since this resource is immutable.

---

Making some experiments on this with Terraform CLI  to better understand the problem:


With `autovacuum` is `on` on the server-side and with the following resource:

```tf
resource "azurerm_postgresql_configuration" "av" {
  name                = "autovacuum"
  resource_group_name = "alper"
  server_name         = "example-psqlserver-alper"
  value               = "on"

  lifecycle {
    prevent_destroy = true
  }
}
```

when I try to run `terraform plan`, no matter I configure `on` or `off`, I get output saying `1 to add, 0 to change, 0 to destroy.` which is weird as you mentioned because I would expect `1 to change` and `0 to change` respectively. 

When I apply `off`, it ends up with `1 added` and I observe the config updated on the server-side, and for the next `terraform plan` or `terraform plan --refresh-only`, it reports properly with no error: `No changes. Your infrastructure matches the configuration.`. It is also the same if I remove all attributes and only leave `id` there before refresh only, just like we do in Terrajet.

So, I am a bit confused at this point about what I am doing differently. Based on my experiments, the following should already be possible:

a. server side config x -> Crossplane MR wants x -> ready&synced
b. server side config x -> Crossplane MR wants y -> properly updated server-side to y -> ready&synced

What will NOT be possible is:

a. server-side config x -> Crossplane MR wants y -> properly updated server-side to y -> ready&synced -> Crossplane MR updated and wants z -> it fails.

**Comment by turkenh** [comment]
> However, there is no case that this resource does not exist, it always imports the server-side default, then you get the problem that there needs to be re-creation because resource already exists. So, any sort of creation is actually re-creation for that resource.

I am not sure about this. My understanding is, there **_could be_** at most one config resource which overrides the default. If there is none overriding the default, then, a new one is created.

**Comment by muvaf** [comment]
@ulucinar Alright, it seems like I used `NameAsIdentifier` during the addition of the configs incorrectly. This is probably why it's importing the resource. I remember us discussing that it should use `IdentifierFromProvider` since defaulting to `metadata.name` doesn't make much sense because the name is from config set of PostgreSQL. @turkenh just tested what happens if we don't write the ID in the tfstate before refresh and it's working as we expected, i.e. returns `NotFound` in the first `Refresh`.

We can re-consider the whole lifecycle thing but it seems like this specific problem can be solved by treating the ID of this resource calculated by Azure.

**Comment by ulucinar** [comment]
Thank you for taking a look at this issue @muvaf & @turkenh. If I summarize my observations: 

- The first `terraform apply -refresh-only -auto-approve -input=false -lock=false -json` that we run in `external.Observe -> Workspace.Refresh` returns success and causes the MR to transition to the `Ready` state because according to ARM, these configuration resources always exist, even if they  have never been "created" before. 
- Then comes the `terraform plan -refresh=false -input=false -lock=false -json` that we run in `external.Observe -> Workspace.Plan` because the resource (always) exists. And if the desired server configuration value is different than the observed configuration value (observed within the `Workspace.Refresh`), then because Terraform provider needs to destroy the resource to change its value, it fails because we have the `prevent_destroy` meta argument.

I think we are on the same page here. As you mentioned, if we remove the `id` attribute from the tfstate we generate, then it should work because in `external.Observe` we will conclude that the resource does not exist and directly run a `terraform apply -auto-approve -input=false -lock=false -json`. However, updates will still fail as expected. 

What I did not understand is that we always place this `id` attribute [here](https://github.com/crossplane-contrib/terrajet/blob/b9fa3b659d50ce40d04023738d53fb6c667856c0/pkg/terraform/files.go#L108) while generating the tfstate. How can we prevent that?

**Comment by muvaf** [comment]
@ulucinar `config.IdentifierFromProvider` has a no-op `GetIDFn`, which means ID ends up being empty and Terraform thinks that the resource is deleted, hence empties the tfstate which is what we use to understand `NotFound`.

**Comment by muvaf** [comment]
By no-op, I mean it just propagates the external name, which is empty in the first run because we disable the name initializer https://github.com/crossplane-contrib/terrajet/blob/main/pkg/config/defaults.go#L45

**Comment by ulucinar** [comment]
Thank you @muvaf for the clarification. After checking your mentioned changes I understood your comment. 

As we discussed offline, I think the current implementation is correct as we compute a valid Azure ID and as an intricacy of these server configuration resources, they exist as long as the associated server resource exists as I mentioned [above](https://github.com/crossplane-contrib/provider-jet-azure/issues/97#issuecomment-979351826). Furthermore, even if as a "side effect" of not using the default external-name initializer we will get the initial update working, once the external-name is set, updates to the configuration value will fail. And the real reason of not allowing Cloud resource destroys during updates is to protect "expensive to reconstruct" resources like relational databases. And these server configuration resources can not be categorized as so. 

So I still think that there's value in allowing developers of Terrajet-based providers to configure this Terraform behavior, via our [resource configuration API](https://github.com/crossplane-contrib/terrajet/blob/b9fa3b659d50ce40d04023738d53fb6c667856c0/pkg/config/resource.go#L189). Such resource configuration is done at compile time and the behavior is fixed at runtime. So provider developers will still be able to protect expensive-to-reconstruct resources like databases at all times but will also have a way to change this behavior for other resources if Terraform has chosen to implement the other way, like in the case with server configurations. Unless explicitly configured otherwise, the default would be to protect resources with the `prevent_destroy` meta-arg. 

As we have decided offline, I will continue with disabling the external-name initializer so that we will have an initial empty id [here](https://github.com/crossplane-contrib/terrajet/blob/b9fa3b659d50ce40d04023738d53fb6c667856c0/pkg/terraform/files.go#L108) for the initial update to succeed. And we will have a limitation of not-being to update server configuration values, which is not the case for Terraform's native server configuration resources like `postgresql_configuration`. We may further discuss the uses cases in a Terrajet issue.

Thank you both for providing valuable input for the discussion. Very much appreciated!
